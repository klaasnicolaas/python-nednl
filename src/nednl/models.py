"""Asynchronous Python client for National Energy Dashboard NL."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Generic, TypeVar

from mashumaro import field_options
from mashumaro.config import BaseConfig
from mashumaro.mixins.dict import DataClassDictMixin
from mashumaro.mixins.orjson import DataClassORJSONMixin

_ResultDataT = TypeVar("_ResultDataT")


@dataclass
class NedDataMixin(DataClassDictMixin):
    """Mixin for dataclasses with a 'data' attribute."""

    # pylint: disable-next=too-few-public-methods
    class Config(BaseConfig):
        """Configuration for mashumaro."""

        serialize_by_alias = True


@dataclass
class BaseResponse(Generic[_ResultDataT], NedDataMixin, DataClassORJSONMixin):
    """Base object representing the API response."""

    data: _ResultDataT = field(metadata=field_options(alias="hydra:member"))
    items: int = field(metadata=field_options(alias="hydra:totalItems"))


@dataclass(slots=True)
class Activity(NedDataMixin, DataClassORJSONMixin):
    """Object representing an activity from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    name: str


@dataclass(slots=True)
class Classification(NedDataMixin, DataClassORJSONMixin):
    """Object representing an classification from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    name: str


@dataclass(slots=True)
class Granularity(NedDataMixin, DataClassORJSONMixin):
    """Object representing an granularity from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    name: str


@dataclass(slots=True)
class GranularityTimezone(NedDataMixin, DataClassORJSONMixin):
    """Object representing an granularity timezone from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    name: str


@dataclass(slots=True)
class Point(NedDataMixin, DataClassORJSONMixin):
    """Object representing an area from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    name: str
    shortname: str = field(metadata=field_options(alias="nameshort"))


@dataclass(slots=True)
class Type(NedDataMixin, DataClassORJSONMixin):
    """Object representing an type from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    name: str
    shortname: str = field(metadata=field_options(alias="nameshort"))


@dataclass(slots=True)
class Utilization(NedDataMixin, DataClassORJSONMixin):
    """Object representing an utilization from National Energy Dashboard NL."""

    id: int  # noqa: A003, RUF100
    capacity: int
    volume: int
    percentage: float
    emission: int
    emission_factor: int = field(metadata=field_options(alias="emissionfactor"))
    valid_from: datetime = field(metadata=field_options(alias="validfrom"))
    valid_to: datetime = field(metadata=field_options(alias="validto"))
    last_update: datetime = field(metadata=field_options(alias="lastupdate"))


@dataclass(slots=True)
class ActivitiesResponse(BaseResponse[list[Activity]]):
    """Object representing an Activities API response."""


@dataclass(slots=True)
class ClassificationsResponse(BaseResponse[list[Classification]]):
    """Object representing an Classifications API response."""


@dataclass(slots=True)
class GranularitiesResponse(BaseResponse[list[Granularity]]):
    """Object representing an Granularities API response."""


@dataclass(slots=True)
class GranularityTimezonesResponse(BaseResponse[list[GranularityTimezone]]):
    """Object representing an GranularityTimezones API response."""


@dataclass(slots=True)
class PointsResponse(BaseResponse[list[Point]]):
    """Object representing an Points API response."""


@dataclass(slots=True)
class TypesResponse(BaseResponse[list[Type]]):
    """Object representing an Types API response."""


@dataclass(slots=True)
class UtilizationsResponse(BaseResponse[list[Utilization]]):
    """Object representing an Utilizations API response."""
