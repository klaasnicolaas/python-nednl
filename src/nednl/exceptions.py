"""Asynchronous Python client for National Energy Dashboard NL."""


class NedNLError(Exception):
    """Generic NED NL exception."""


class NedNLConnectionError(NedNLError):
    """NED NL connection exception."""


class NedNLAuthenticationError(NedNLError):
    """NED NL authentication exception."""
