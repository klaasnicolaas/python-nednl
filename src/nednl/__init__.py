"""Asynchronous Python client for National Energy Dashboard NL."""

from .exceptions import NedNLAuthenticationError, NedNLConnectionError, NedNLError
from .models import (
    Activity,
    Classification,
    Granularity,
    GranularityTimezone,
    Point,
    Type,
    Utilization,
)
from .nednl import NedNL

__all__ = [
    "Activity",
    "Classification",
    "Granularity",
    "GranularityTimezone",
    "NedNL",
    "NedNLAuthenticationError",
    "NedNLConnectionError",
    "NedNLError",
    "Point",
    "Type",
    "Utilization",
]
