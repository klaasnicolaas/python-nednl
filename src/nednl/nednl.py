"""Asynchronous Python client for National Energy Dashboard NL."""

from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self

from aiohttp import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import NedNLAuthenticationError, NedNLConnectionError, NedNLError
from .models import (
    ActivitiesResponse,
    Activity,
    Classification,
    ClassificationsResponse,
    GranularitiesResponse,
    Granularity,
    GranularityTimezone,
    GranularityTimezonesResponse,
    Point,
    PointsResponse,
    Type,
    TypesResponse,
    Utilization,
    UtilizationsResponse,
)

VERSION = metadata.version(__package__)


@dataclass
class NedNL:
    """Main class for handling data fetching from National Energy Dashboard NL."""

    api_key: str
    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the National Energy Dashboard NL API.

        Args:
        ----
            uri: Request URI, without '/api/', for example, 'status'.
            method: HTTP method to use.
            params: Extra options to improve or limit the response.

        Returns:
        -------
            The response data from the API.

        Raises:
        ------
            NedNLAuthenticationError: If no API key is provided.
            NedNLConnectionError: If an error occurs while connecting to the API.
            NedNLError: If an unexpected response is received from the API.

        """
        url = URL.build(scheme="https", host="api.ned.nl", path="/v1/").join(URL(uri))

        if self.api_key is None or self.api_key == "":
            msg = "No API key provided."
            raise NedNLAuthenticationError(msg)

        headers = {
            "Accept": "application/ld+json",
            "User-Agent": f"python-nednl/{VERSION}",
            "X-AUTH-TOKEN": self.api_key,
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except TimeoutError as exception:
            msg = "Timeout occurred while connecting to NED NL API."
            raise NedNLConnectionError(msg) from exception
        except ClientResponseError as exception:
            if exception.status == 403:
                msg = "Invalid or expired API key provided."
                raise NedNLAuthenticationError(msg) from exception
            msg = "Error occurred while communicating with NED NL API."
            raise NedNLConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with NED NL API."
            raise NedNLConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        text = await response.text()
        if "application/ld+json" not in content_type:
            msg = "Unexpected content type response from NED NL API."
            raise NedNLError(
                msg,
                {"Content-Type": content_type, "Response": text},
            )

        return text

    async def all_activities(self) -> list[Activity]:
        """Get list of all activities.

        Returns
        -------
            List of all activities.

        """
        response = await self._request("activities")
        return ActivitiesResponse.from_json(response).data

    async def all_classifications(self) -> list[Classification]:
        """Get list of all classifications.

        Returns
        -------
            List of all classifications.

        """
        response = await self._request("classifications")
        return ClassificationsResponse.from_json(response).data

    async def all_granularities(self) -> list[Granularity]:
        """Get list of all granularities.

        Returns
        -------
            List of all granularities.

        """
        response = await self._request("granularities")
        return GranularitiesResponse.from_json(response).data

    async def all_granularity_timezones(self) -> list[GranularityTimezone]:
        """Get list of all granularity timezones.

        Returns
        -------
            List of all granularity timezones.

        """
        response = await self._request("granularity_time_zones")
        return GranularityTimezonesResponse.from_json(response).data

    async def all_points(self) -> list[Point]:
        """Get list of all area points.

        Returns
        -------
            List of all area points.

        """
        response = await self._request("points")
        return PointsResponse.from_json(response).data

    async def all_types(self) -> list[Type]:
        """Get list of all types.

        Returns
        -------
            List of all types.

        """
        response = await self._request("types", params={"itemsPerPage": 100})
        return TypesResponse.from_json(response).data

    async def utilization(  # noqa: PLR0913, pylint: disable=too-many-arguments
        self,
        point_id: int,
        type_id: int,
        granularity_id: int,
        granularity_timezone_id: int,
        classification_id: int,
        activity_id: int,
        start_date: str,
        end_date: str,
    ) -> list[Utilization]:
        """Get utilization data for a specific point, granularity, and time.

        Args:
        ----
            point_id: The ID of the point.
            type_id: The ID of the type.
            granularity_id: The ID of the granularity.
            granularity_timezone_id: The ID of the granularity timezone.
            classification_id: The ID of the classification.
            activity_id: The ID of the activity.
            start_date: The start date of the data.
            end_date: The end date of the data.

        Returns:
        -------
            Utilization data for the specific point, granularity, and time.

        """
        response = await self._request(
            "utilizations",
            params={
                "point": point_id,
                "type": type_id,
                "granularity": granularity_id,
                "granularitytimezone": granularity_timezone_id,
                "classification": classification_id,
                "activity": activity_id,
                "validfrom[strictly_after]": start_date,
                "validfrom[strictly_before]": end_date,
            },
        )
        return UtilizationsResponse.from_json(response).data

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The National Energy Dashboard NL object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
