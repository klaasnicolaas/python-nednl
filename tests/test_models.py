"""Test the models for National Energy Dashboard NL."""

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from nednl import NedNL

from . import load_fixtures


async def test_activities_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    nednl_client: NedNL,
) -> None:
    """Test activities data is handled correctly."""
    aresponses.add(
        "api.ned.nl",
        "/v1/activities",
        "GET",
        aresponses.Response(
            status=200,
            content_type="application/ld+json",
            body=load_fixtures("activities.json"),
        ),
    )
    response = await nednl_client.all_activities()
    assert response == snapshot


async def test_classifications_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    nednl_client: NedNL,
) -> None:
    """Test classifications data is handled correctly."""
    aresponses.add(
        "api.ned.nl",
        "/v1/classifications",
        "GET",
        aresponses.Response(
            status=200,
            content_type="application/ld+json",
            body=load_fixtures("classifications.json"),
        ),
    )
    response = await nednl_client.all_classifications()
    assert response == snapshot


async def test_granularities_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    nednl_client: NedNL,
) -> None:
    """Test granularities data is handled correctly."""
    aresponses.add(
        "api.ned.nl",
        "/v1/granularities",
        "GET",
        aresponses.Response(
            status=200,
            content_type="application/ld+json",
            body=load_fixtures("granularities.json"),
        ),
    )
    response = await nednl_client.all_granularities()
    assert response == snapshot


async def test_points_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    nednl_client: NedNL,
) -> None:
    """Test points data is handled correctly."""
    aresponses.add(
        "api.ned.nl",
        "/v1/points",
        "GET",
        aresponses.Response(
            status=200,
            content_type="application/ld+json",
            body=load_fixtures("points.json"),
        ),
    )
    response = await nednl_client.all_points()
    assert response == snapshot


async def test_types_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    nednl_client: NedNL,
) -> None:
    """Test types data is handled correctly."""
    aresponses.add(
        "api.ned.nl",
        "/v1/types",
        "GET",
        aresponses.Response(
            status=200,
            content_type="application/ld+json",
            body=load_fixtures("types.json"),
        ),
    )
    response = await nednl_client.all_types()
    assert response == snapshot
