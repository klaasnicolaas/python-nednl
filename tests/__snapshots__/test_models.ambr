# serializer version: 1
# name: test_granularities_data
  list([
    Granularity(id=3, name='10Min'),
    Granularity(id=4, name='15Min'),
    Granularity(id=5, name='Hour'),
    Granularity(id=6, name='Day'),
    Granularity(id=7, name='Month'),
    Granularity(id=8, name='Year'),
  ])
# ---
# name: test_points_data
  list([
    Point(id=0, name='Nederland', nameshort='NL'),
    Point(id=1, name='Groningen', nameshort='Gr'),
    Point(id=2, name='Friesland', nameshort='Fr'),
    Point(id=3, name='Drenthe', nameshort='Dr'),
    Point(id=4, name='Overijssel', nameshort='Ov'),
    Point(id=5, name='Flevoland', nameshort='FL'),
    Point(id=6, name='Gelderland', nameshort='Ge'),
    Point(id=7, name='Utrecht', nameshort='Ut'),
    Point(id=8, name='Noord-Holland', nameshort='No'),
    Point(id=9, name='Zuid-Holland', nameshort='Zu'),
    Point(id=10, name='Zeeland', nameshort='Ze'),
    Point(id=11, name='Noord-Brabant', nameshort='Nb'),
    Point(id=12, name='Limburg', nameshort='Li'),
    Point(id=13, name='Eilanden', nameshort='Ei'),
    Point(id=28, name='Windpark Luchterduinen', nameshort='WPlu'),
    Point(id=29, name='Windpark Princes Amalia', nameshort='WPam'),
    Point(id=30, name='Windpark Egmond aan Zee', nameshort='WPoz'),
    Point(id=31, name='Windpark Gemini', nameshort='WPGe'),
    Point(id=32, name='Windpark Borselle', nameshort='WPbo'),
    Point(id=33, name='Windpark Borselle I&II', nameshort='WPb1'),
    Point(id=34, name='Windpark Borselle III&IV', nameshort='Wpb3'),
    Point(id=35, name='Windpark Hollandse Kust Zuid', nameshort='WPkz'),
    Point(id=36, name='Windpark Hollandse Kust Noord', nameshort='WPkn'),
  ])
# ---
